app/scripts/lib/experiments/base.js

```diff
@@ -0,0 +1,164 @@
+     */
+    start: function (name, options) {
+      // all experiments require these options
+      if (! name || ! options || ! options.window || ! options.metrics || ! options.able || ! options.user) {
+        return false;
+      }
+
+      this.window = options.window;
+      this.able = options.able;
+      this.metrics = options.metrics;
+      this.user = options.user;
+      this.extraAbleOptions = options.extraAbleOptions || {};
+
+      var abData = {
+        isMetricsEnabledValue: this.metrics.isCollectionEnabled(),
+        uniqueUserId: this.user.get('uniqueUserId'),
+        // the window parameter will override any ab testing features
+        forceExperimentGroup: Url.searchParam(FORCE_GROUP_TYPE, this.window.location.search)
+      };
+
+      abData = _.extend(abData, this.extraAbleOptions);
+
+      this._name = name;
+      this._groupType = this.able.choose(this._name, abData) || null;
+      if (! this._groupType) {
+        return false;
+      }
+
+      this._loggingNamespace = 'experiment.' + this._groupType + '.' + this._name + '.';
+      this._storageNamespace = 'experiment.' + this._name;
+      this._initialized = true;
+      this.saveState('enrolled');
+
+      return true;
+    },
+    /**
+     * Checks the experiment name
+     *
+     * @param name
+     * @returns {boolean}
+     */
+    is: function (name) {
+      return !! name && name === this._name;
+    },
+    /**
+     * Checks the experiment group
+     *
+     * @param groupType
+     * @returns {boolean}
+     */
+    isGroup: function (groupType) {
+      return !! groupType && groupType === this._groupType;
+    },
+    /**
+     * Returns the extra context of an experiment
+     *
+     * @returns {Object}
+     */
+    getContext: function () {
+      // context should be extended
+      return {};
+    },
+    /**
+     * Series of actions that an experiment can perform
+     *
+     * @param action
+     * @param options
+     */
+    performAction: function (action, options) {
+      // actions should be extended
+    },
+    /**
+     * Log an experiment event
+     *
+     * @param event
+     */
+    logEvent: function (event) {
+      if (this._initialized && event) {
+        this.metrics.logEvent(this._loggingNamespace + event);
+      }
+    },
+    /**
+     * Save that some experiment state occurred and
+     * log an event if it was the first time it happened
+     *
+     * @param state
+     */
+    saveState: function (state) {
+      if (! state) {
+        return;
+      }
+
+      var store = JSON.parse(storage.get(this._storageNamespace)) || {};
+      if (! store[state]) {
+        this.logEvent(state);
+      }
+      store[state] = true;
+      storage.set(this._storageNamespace, JSON.stringify(store));
+    },
+    /**
+     * Check if the experiment had some state occur
+     *
+     * @param state
+     * @returns {boolean}
+     */
+    hasState: function (state) {
```

Comment:
There is `saveState` and `hasState`, but no `getState`. Where is the state used other than `hasState`, and if it's only here, is it sufficient just to save a simple flag into storage instead of a full object?